library-udf/src/main/java/org/apache/iotdb/library/dlearn/UDTFAR.java

```diff
@@ -0,0 +1,132 @@
+ * under the License.
+ */
+
+package org.apache.iotdb.library.dlearn;
+
+import org.apache.iotdb.library.util.Util;
+import org.apache.iotdb.udf.api.UDTF;
+import org.apache.iotdb.udf.api.access.Row;
+import org.apache.iotdb.udf.api.collector.PointCollector;
+import org.apache.iotdb.udf.api.customizer.config.UDTFConfigurations;
+import org.apache.iotdb.udf.api.customizer.parameter.UDFParameters;
+import org.apache.iotdb.udf.api.customizer.strategy.RowByRowAccessStrategy;
+import org.apache.iotdb.udf.api.type.Type;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class UDTFAR implements UDTF {
+  private int p;
+  private long interval;
+  private List<Long> timeWindow = new ArrayList<>();
+  private List<Double> valueWindow = new ArrayList<>();
+
+  @Override
+  public void beforeStart(UDFParameters udfParameters, UDTFConfigurations udtfConfigurations)
+      throws Exception {
+    udtfConfigurations
+        .setAccessStrategy(new RowByRowAccessStrategy())
+        .setOutputDataType(udfParameters.getDataType(0));
+    this.p = udfParameters.getIntOrDefault("p", 1);
+    udtfConfigurations.setAccessStrategy(new RowByRowAccessStrategy());
+    udtfConfigurations.setOutputDataType(Type.DOUBLE);
+  }
+
+  @Override
+  public void transform(Row row, PointCollector collector) throws Exception {
+    if (!row.isNull(0)) {
+      timeWindow.add(row.getTime());
+      valueWindow.add(Util.getValueAsDouble(row));
+    }
+  }
+
+  @Override
+  public void terminate(PointCollector collector) throws Exception {
+    int length = timeWindow.size();
+    if (length <= this.p) {
+      collector.putString(1, "illegal input");
+      return;
+    }
+
+    int count = 0;
+    long maxFreqInterval = timeWindow.get(1) - timeWindow.get(0);
+    for (int i = 2; i < timeWindow.size(); i++) {
+      if (maxFreqInterval == timeWindow.get(i) - timeWindow.get(i - 1)) count++;
+      else {
+        count--;
+        if (count == 0) {
+          maxFreqInterval = timeWindow.get(i) - timeWindow.get(i - 1);
+          count++;
+        }
+      }
+    }
+    this.interval = maxFreqInterval;
+
+    List<Long> imputedTimeWindow = new ArrayList<>();
+    List<Double> imputedValueWindow = new ArrayList<>();
+
+    imputedTimeWindow.add(timeWindow.get(0));
+    imputedValueWindow.add(valueWindow.get(0));
+    for (int i = 1; i < timeWindow.size(); i++) {
+      if (timeWindow.get(i) - timeWindow.get(i - 1) > interval) {
+        int gap = (int) ((timeWindow.get(i) - timeWindow.get(i - 1)) / interval);
+        double step = (valueWindow.get(i) - valueWindow.get(i - 1)) / gap;
+        for (int j = 1; j < gap; j++) {
+          imputedTimeWindow.add(timeWindow.get(i - 1) + j * interval);
+          imputedValueWindow.add(valueWindow.get(i - 1) + j * step);
+        }
+      }
+      imputedTimeWindow.add(timeWindow.get(i));
+      imputedValueWindow.add(valueWindow.get(i));
+    }
+    System.out.println("length with imputation:" + imputedTimeWindow.size());
```

Comment:
You could init a Logger and use Logger.debug or Logger.info if you want to print something.